### PR TITLE
Do async locking correctly

### LINKED ISSUE #6012: RecoverableTextAndVersion using lock improperly.
The GetValueAsync API should be using the semaphore slim asynchronously.  Instead, it is using it synchronously, across an await call.  Maybe we need an analyzer here.

This is related to connect issue/devdiv work item 132147.

### PR DESCRIPTION
**Pull request by mattwar**
Fixes #6012 

This potentially fixes problem reported in devdiv work item 132147.

### CODE REVIEW

File: build/ToolsetPackages/project.lock.json [hunk 1]
```diff
@@ -30,7 +30,7 @@
         }
       },
       "Microsoft.DiaSymReader.Native/1.2.0-rc": {},
-      "Microsoft.Net.Compilers/1.1.0-beta1": {},
+      "Microsoft.Net.Compilers/1.1.0-beta1-20150727-01": {},
```
Comment (Pilchie) [inline build/ToolsetPackages/project.lock.json]: This file should probably not be included.

**Comment by Pilchie** [comment]
LGTM apart from the lock file.

**Comment by Pilchie** [comment]
Also - we should consider whether this is warrants being cherry-picked to the stabilization branch

File: build/ToolsetPackages/project.lock.json (hunk 1 above)
Comment (mattwar) [inline build/ToolsetPackages/project.lock.json]: I was told explicitly, if my building Roslyn causes changes to project.lock.json, I was supposed to check them in.

**Pull request merged**